lmp/infer/_top_p.py

```diff
@@ -0,0 +1,294 @@
+r"""Top ``P`` inference method."""
+
+import argparse
+from typing import ClassVar, Dict, Optional
+
+import torch
+
+from lmp.infer._base import BaseInfer
+from lmp.model import BaseModel
+from lmp.tknzr import BaseTknzr
+
+
+class TopPInfer(BaseInfer):
+    r"""Top ``P`` inference method.
+
+    Use indice with the top ``P'`` highest probability as possible next token
+    id, then randomly choose ``1`` index out of ``P'`` as next token id
+    prediction.
+    Top-p sampling, also called nucleus sampling is similar to top-k sampling
+    where k changes at every step tp cover ``P`` probability mass.
+    It is a non-greedy algorithm since the best prediction is not always
+    choosen, but it provide dynamic of generation result (because of
+    randomness, obviously).
+
+    For comment throughout this class and its subclasses, we use ``P'`` to
+    denote the number of candidate token ids with highest probabilities that
+    cover ``P`` probability mass.
+
+    Attributes
+    ==========
+    infer_name: ClassVar[str]
+        Inference method name is ``top-p``.
+        Used for command line argument parsing.
+    """
+    infer_name: ClassVar[str] = 'top-p'
+    def __init__(self, p: float, max_seq_len: int, **kwargs: Optional[Dict]):
+        super().__init__(max_seq_len=max_seq_len)
+        if not isinstance(p, float):
+            raise TypeError('`p` must be an instance of `float`.')
+        if not p <= 1.0:
+            raise ValueError('`p` must satisfy `p < 0`.')
+
+        self.p = p
+
+    @torch.no_grad()
+    def gen(
+            self,
+            model: BaseModel,
+            tknzr: BaseTknzr,
+            txt: str,
+    ) -> str:
+        r"""Generate text conditional on text segment.
+
+        Top ``P`` inference algorithm is structured as follow:
+
+        #. Encode input text as ``1`` sample batch.
+           (shape: ``(1, S')``)
+        #. Remove ``[eos]`` token since model is not trained to predict tokens
+           after seeing ``[eos]``.
+           (shape: ``(1, S'-1)`` or ``(1, S)`` where ``S'-1 = S``)
+        #. Truncate text to satisfy maximum sequence constraint.
+           (shape: ``(1, S)`` or ``(1, max_seq_len)``)
+        #. Use for-loop to generate sequence of token ids.
+
+            #. Use ``model.pred()`` to get next token ids probability
+               distribution.
+               (shape: ``(1, S, V)``)
+            #. Get the last next token id probability distribution.
+               (shape: ``(1, V)``)
+            #. Sort the probability distribution in descending order.
+               (shape: ``(1, V)``)
+            #. Get the P' highest probability distribution and their
```

Comment:
redundant quotation mark in `P'`

Reply:
remove "where P' changes at every time step to
               cover P probability mass."